src/scanspec/core.py

```diff
@@ -37,140 +31,159 @@
-                # Build the alias of the field
-                alias = to_pascal_case(constructor.__name__ + sub.__name__)
-                # object_deserialization uses get_type_hints, but the constructor
-                # return type is stringified and the class not defined yet,
-                # so it must be assigned manually
-                constructor.__annotations__["return"] = with_params(sub)
-                # Use object_deserialization to wrap constructor as deserializer
-                deserialization = object_deserialization(
-                    constructor, type_name_factory_for_subclass(sub)
-                )
-                # Add constructor tagged field with its conversion
-                annotations[alias] = Tagged[with_params(sub)]  # type: ignore
-                namespace[alias] = Tagged(conversion(deserialization=deserialization))
-        # Create the deserialization tagged union class
-        tagged_union = new_class(
-            cls.__name__, tagged_union_bases, exec_body=lambda ns: ns.update(namespace)
-        )
-        return Conversion(
-            lambda obj: get_tagged(obj)[1],
-            source=with_params(tagged_union),
-            target=with_params(cls),
+        my_sum = Add(IntLiteral(5), Subtract(IntLiteral(10), IntLiteral(2)))
+        assert my_sum.calculate() == 13
+
+        assert my_sum == parse_obj_as(
+            Expression,
+            {
+                "type": "Add",
+                "left": {"type": "IntLiteral", "value": 5},
+                "right": {
+                    "type": "Subtract",
+                    "left": {"type": "IntLiteral", "value": 10},
+                    "right": {"type": "IntLiteral", "value": 2},
+                },
+            },
         )
 
-    deserializer(lazy=deserialization, target=cls)
-    serializer(lazy=serialization, source=cls)
-    return cls
+    Args:
+        super_cls: The superclass of the union, Expression in the above example
+        discriminator: The discriminator that will be inserted into the
+            serialized documents for type determination. Defaults to "type".
+        config: A pydantic config class to be inserted into all
+            subclasses. Defaults to None.
+
+    Returns:
+        Union[Type, Callable[[Type], Type]]: A decorator that adds the necessary
+            functionality to a class.
+    """
+
+    def wrap(cls):
+        return _discriminated_union_of_subclasses(cls, discriminator, config)
+
+    # Work out if the call was @discriminated_union_of_subclasses or
+    # @discriminated_union_of_subclasses(...)
+    if super_cls is None:
+        return wrap
+    else:
+        return wrap(super_cls)
+
+
+def _discriminated_union_of_subclasses(
+    super_cls: Type,
+    discriminator: str,
+    config: Optional[Type[BaseConfig]] = None,
+) -> Union[Type, Callable[[Type], Type]]:
+
+    super_cls.__ref_classes__ = set()
+    super_cls.__model__ = None
```

Comment:
Python stdlib reserves the right to use any dunder methods for its own methods. Suggest we keep a couple of global dicts around for this instead of attaching to the cls

Reply:
It looks like you only use these in `__validate__`, so they could live anywhere if this is the case

Reply:
Good point, that works, although I'm unsure if using globals where we don't absolutely need to is any better than using dunders. It would be far better if this information were attached to the class it was about. Can we just turn it into `super_cls._models`?

Reply:
Have renamed to `super_cls._ref_classes` and `super_cls._models`